File: backend/analysis_engine.py
from typing import List, Dict

# Simple placeholder: later you can load this from a DB of TT laws
OFFENCE_ELEMENTS = {
    "robbery": [
        "theft of property",
        "use or threat of force",
        "without consent",
        "intent to permanently deprive",
    ],
    "firearm": [
        "possession of firearm",
        "without lawful authority",
        "knowledge of possession",
    ],
    "dangerous_drugs": [
        "possession of dangerous drug",
        "knowledge of possession",
        "without lawful excuse",
    ],
}


def analyze_evidence(evidence_summary: str, offence: str) -> Dict:
    elements = OFFENCE_ELEMENTS.get(offence.lower(), [])
    analysis_results = []
    recommendations = []

    for element in elements:
        keywords = element.split()
        score = 0
        for word in keywords:
            if word.lower() in evidence_summary.lower():
                score += 1

        if score >= len(keywords) - 1:
            strength = "strong"
        elif score >= 1:
            strength = "medium"
        else:
            strength = "weak"

        analysis_results.append(
            {
                "element": element,
                "strength": strength,
                "details": f"Evidence for '{element}' appears {strength} based on current summary.",
            }
        )

        if strength == "weak":
            recommendations.append(
                f"Gather more specific evidence to prove: {element}."
            )
        elif strength == "medium":
            recommendations.append(
                f"Clarify and corroborate evidence relating to: {element}."
            )

    overall_strength = _compute_overall_strength(analysis_results)

    return {
        "offence": offence,
        "overall_strength": overall_strength,
        "elements_analysis": analysis_results,
        "recommendations": recommendations,
    }


def _compute_overall_strength(elements_analysis: List[Dict]) -> str:
    strengths = [e["strength"] for e in elements_analysis]
    if not strengths:
        return "unknown"

    if all(s == "strong" for s in strengths):
        return "very strong case"
    if "weak" in strengths and "strong" not in strengths:
        return "weak case"
    if "weak" in strengths and "strong" in strengths:
        return "mixed case"
    return "moderate case"

File: web/src/lib/analysis-engine.ts
import type { AnalysisResult } from "./types";

const OFFENCE_ELEMENTS: Record<string, string[]> = {
  robbery: [
    "theft of property",
    "use or threat of force",
    "without consent",
    "intent to permanently deprive",
  ],
  firearm: [
    "possession of firearm",
    "without lawful authority",
    "knowledge of possession",
  ],
  dangerous_drugs: [
    "possession of dangerous drug",
    "knowledge of possession",
    "without lawful excuse",
  ],
};

function computeOverallStrength(
  elements: { strength: string }[]
): string {
  const strengths = elements.map((e) => e.strength);
  if (strengths.length === 0) return "unknown";
  if (strengths.every((s) => s === "strong")) return "very strong case";
  if (strengths.includes("weak") && !strengths.includes("strong"))
    return "weak case";
  if (strengths.includes("weak") && strengths.includes("strong"))
    return "mixed case";
  return "moderate case";
}

export function analyzeEvidence(
  evidenceSummary: string,
  offence: string
): AnalysisResult {
  const elements = OFFENCE_ELEMENTS[offence.toLowerCase()] ?? [];
  const analysisResults = [];
  const recommendations: string[] = [];

  for (const element of elements) {
    const keywords = element.split(" ");
    let score = 0;
    for (const word of keywords) {
      if (evidenceSummary.toLowerCase().includes(word.toLowerCase())) {
        score += 1;
      }
    }

    let strength: "strong" | "medium" | "weak";
    if (score >= keywords.length - 1) strength = "strong";
    else if (score >= 1) strength = "medium";
    else strength = "weak";

    analysisResults.push({
      element,
      strength,
      details: `Evidence for '${element}' appears ${strength} based on current summary.`,
    });

    if (strength === "weak") {
      recommendations.push(
        `Gather more specific evidence to prove: ${element}.`
      );
    } else if (strength === "medium") {
      recommendations.push(
        `Clarify and corroborate evidence relating to: ${element}.`
      );
    }
  }

  return {
    offence,
    overall_strength: computeOverallStrength(analysisResults),
    elements_analysis: analysisResults,
    recommendations,
  };
}

export function formatAnalysisText(result: AnalysisResult): string {
  return result.elements_analysis
    .map(
      (e) =>
        `${e.element}: ${e.strength.toUpperCase()} — ${e.details}`
    )
    .join("\n\n");
}

export function buildCrossExamination(result: AnalysisResult): string {
  const weakElements = result.elements_analysis.filter(
    (e) => e.strength === "weak" || e.strength === "medium"
  );

  if (weakElements.length === 0) {
    return "Defence may challenge chain of custody, witness credibility, and whether each statutory element is proved beyond reasonable doubt.";
  }

  return weakElements
    .map(
      (e, i) =>
        `Q${i + 1}: You agree the prosecution must prove "${e.element}" beyond reasonable doubt — can you point to the specific evidence that does so?`
    )
    .join("\n\n");
}
